code/modules/food_and_drinks/kitchen_machinery/kitchen_machine.dm

```diff
@@ -361,97 +318,188 @@
 	return
 
-/obj/machinery/kitchen_machine/examine(mob/user, infix = "", suffix = "")
+/obj/machinery/kitchen_machine/examine(mob/user)
 	. = ..()
 	var/dat = format_content_descs()
 	if(dat)
 		. += "It contains: <BR>[dat]"
+
+/obj/machinery/kitchen_machine/attack_hand(mob/user)
+	if(stat & (BROKEN|NOPOWER) || panel_open || !anchored)
+		return
+
+	ui_interact(user)
+
+/obj/machinery/kitchen_machine/ui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = TRUE, datum/tgui/master_ui = null, datum/ui_state/state = GLOB.default_state)
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "KitchenMachine",  name, 400, 300, master_ui, state)
+		ui.open()
+
+/obj/machinery/kitchen_machine/ui_data(mob/user)
+	var/list/data = list()
+	data["name"] = name
+	data["operating"] = operating
+	data["inactive"] = FALSE
+	data["no_eject"] = FALSE
+	data["tooltip"] = ""
+	if(dirty >= 100)
+		data["inactive"] = TRUE
+		data["tooltip"] = "It's too dirty."
+	else if(!length(contents))
+		data["inactive"] = TRUE
+		data["no_eject"] = TRUE
+		data["tooltip"] = "Theres no contents."
+
+	var/list/items_counts = list()
+	var/list/name_overrides = list()
+	for(var/obj/O in contents)
+		var/display_name = O.name
+		if(istype(O,/obj/item/reagent_containers/food/snacks/egg)) // God I fuckin hate this
+			name_overrides[display_name] = "egg"
+			if(items_counts[display_name])
+				name_overrides[display_name] = "eggs"
+		if(istype(O,/obj/item/reagent_containers/food/snacks/tofu))
+			name_overrides[display_name] = "tofu chunk"
+			if(items_counts[display_name])
+				name_overrides[display_name] = "tofu chunks"
+		if(istype(O,/obj/item/reagent_containers/food/snacks/meat)) //any meat
+			name_overrides[display_name] = "slab of meat"
+			if(items_counts[display_name])
+				name_overrides[display_name] = "slabs of meat"
+		if(istype(O,/obj/item/reagent_containers/food/snacks/donkpocket))
+			display_name = "Turnovers"
+			name_overrides[display_name] = "turnover"
+			if(items_counts[display_name])
+				name_overrides[display_name] = "turnovers"
+		if(istype(O,/obj/item/reagent_containers/food/snacks/carpmeat))
+			name_overrides[display_name] = "fillet of meat"
+			if(items_counts[display_name])
+				name_overrides[display_name] = "fillets of meat"
+		items_counts[display_name]++
+
+	data["ingredients"] = list()
+	for(var/food in items_counts)
+		var/N = items_counts[food]
+		var/units
+		if(!(food in name_overrides))
+			units = "[lowertext(food)]"
+		else
+			units = "[name_overrides[food]]"
+
+		var/list/data_pr = list(
+			name = capitalize(food),
+			amount = N,
+			units = units
+		)
+
+		data["ingredients"] += list(data_pr)
+
+	for(var/datum/reagent/R in reagents.reagent_list)
+		var/display_name = R.name
+		if(R.id == "capsaicin")
+			display_name = "Hotsauce"
+		if(R.id == "frostoil")
+			display_name = "Coldsauce"
+
+		var/unitamt = "unit"
+		if(R.volume > 1)
```

Comment:
Why not just handle this in the TGUI?

Reply:
because you will end up with stuff like "slabs of meats" and "tofu chunkss", reagents gotta have their own unit that isnt that

Reply:
I mean just for a unit vs units. you should just be able to do `amount > 1 ? "s" : ""` or something

Reply:
Yeah, I would only be able to do that inside of dm because they are passed to the same "ingredients" list

Reply:
alright that's fine then